geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/executor/cluster/ClusterExecutor.java

```diff
@@ -0,0 +1,196 @@
+      entry.add(((index + 1) * REDIS_SLOTS_PER_BUCKET) - 1);
+      entry.add(Arrays.asList(primaryHostAndPort.getLeft(), primaryHostAndPort.getRight()));
+
+      slots.add(entry);
+      index++;
+    }
+
+    return RedisResponse.array(slots);
+  }
+
+  @SuppressWarnings("unchecked")
+  private List<MemberBuckets> getMemberBuckets(
+      ExecutionHandlerContext ctx) {
+    Region<RedisKey, RedisData> dataRegion = ctx.getRegionProvider().getDataRegion();
+
+    // Really only need this in situations where the cluster is empty and no data has been
+    // added yet.
+    PartitionRegionHelper.assignBucketsToPartitions(dataRegion);
+
+    Set<DistributedMember> membersWithDataRegion = new HashSet<>();
+    for (PartitionMemberInfo memberInfo : getRegionMembers(ctx)) {
+      membersWithDataRegion.add(memberInfo.getDistributedMember());
+    }
+
+    ResultCollector<MemberBuckets, List<MemberBuckets>> resultCollector =
+        FunctionService.onMembers(membersWithDataRegion).execute(BucketInfoRetrievalFunction.ID);
+
+    return resultCollector.getResult();
+  }
+
+  /**
+   * The format being produced is something like this:
+   *
+   * <pre>
+   * 67ed2db8d677e59ec4a4cefb06858cf2a1a89fa1 127.0.0.1:30002@31002 master - 0 1426238316232 2 connected 5461-10922
+   * 292f8b365bb7edb5e285caf0b7e6ddc7265d2f4f 127.0.0.1:30003@31003 master - 0 1426238318243 3 connected 10923-16383
+   * e7d1eecce10fd6bb5eb35b9f99a514335d9ba9ca 127.0.0.1:30001@31001 myself,master - 0 0 1 connected 0-5460
+   * </pre>
+   *
+   * Note that there are no 'slave' entries since Geode does not host all secondary data apart from
+   * primary as redis does. The cluster port is provided only for consistency with the format
+   * of the output.
+   */
+  private RedisResponse getNodes(ExecutionHandlerContext ctx) {
+    List<MemberBuckets> memberBuckets = getMemberBuckets(ctx);
+    String memberId = ctx.getMemberName();
+
+    StringBuilder response = new StringBuilder();
+    for (MemberBuckets m : memberBuckets) {
+      response.append(String.format("%s %s:%d@%d master",
+          m.getMemberId(), m.getHostAddress(), m.getPort(), m.getPort()));
+
+      if (m.getMemberId().equals(memberId)) {
+        response.append(",myself");
+      }
+      response.append(" - 0 0 1 connected");
+
+      for (Integer index : m.getPrimaryBucketIds()) {
+        response.append(" ");
+        response.append(index * REDIS_SLOTS_PER_BUCKET);
+        response.append("-");
+        response.append(((index + 1) * REDIS_SLOTS_PER_BUCKET) - 1);
+      }
+
+      response.append("\n");
+    }
+
+    return RedisResponse.bulkString(response.toString());
+  }
+
+
+  private RedisResponse getInfo(ExecutionHandlerContext ctx) {
+    int memberCount = getRegionMembers(ctx).size();
+
+    return RedisResponse.bulkString(
+        "cluster_state:ok\r\n"
+            + "cluster_slots_assigned:" + REDIS_SLOTS + "\r\n"
+            + "cluster_slots_ok:" + REDIS_SLOTS + "\r\n"
```

Comment:
At some point we may actually have buckets that are offline or missing. I don't know if there is any value in computing these numbers instead of hardcoding them. Maybe not right now.

Reply:
I think implementing `MOVED` will introduce a framework to track this detail better. At which point we can improve this.

Reply:
This optimization should include some way of sharing hostname and port information for all the Redis endpoints. Perhaps it makes sense for this to go into some Redis metadata replicated region.